drivers/wifi/infineon/airoc_whd_hal.c

```diff
@@ -0,0 +1,474 @@
+	whd_sdio_config.oob_config = oob_config;
+#endif
+
+	if (airoc_wifi_power_on(dev)) {
+		LOG_ERR("airoc_wifi_power_on retuens fail");
+		return -ENODEV;
+	}
+
+	if (!device_is_ready(config->sdhc_dev)) {
+		LOG_ERR("SDHC device is not ready");
+		return -ENODEV;
+	}
+
+	ret = sd_init(config->sdhc_dev, &data->card);
+	if (ret) {
+		return ret;
+	}
+
+	cy_rslt_t result = whd_init(&data->whd_drv, &init_config_default, &resource_ops,
+				    &buffer_if_default, &netif_if_default);
+	if (result == CY_RSLT_SUCCESS) {
+		result = whd_bus_sdio_attach(data->whd_drv, &whd_sdio_config,
+					     (whd_sdio_t)&data->card);
+
+		if (result == CY_RSLT_SUCCESS) {
+			result = whd_wifi_on(data->whd_drv, interface);
+		}
+
+		if (result != CY_RSLT_SUCCESS) {
+			whd_deinit(*interface);
+		}
+	}
+	return result;
+}
+
+/*
+ * Implement SDIO CMD52/53 wrappers
+ */
+whd_result_t whd_bus_sdio_cmd52(whd_driver_t whd_driver, whd_bus_transfer_direction_t direction,
+				whd_bus_function_t function, uint32_t address, uint8_t value,
+				sdio_response_needed_t response_expected, uint8_t *response)
+{
+	whd_result_t result;
+	struct sd_card *sd = (struct sd_card *)whd_driver->bus_priv->sdio_obj;
+	sdio_cmd_argument_t arg = {0};
+	struct sdhc_command cmd = {.opcode = SDIO_RW_DIRECT,
+				   .response_type = SD_RSP_TYPE_R5,
+				   .timeout_ms = CONFIG_SD_CMD_TIMEOUT};
+
+	arg.cmd52.function_number = (uint32_t)(function & BUS_FUNCTION_MASK);
+	arg.cmd52.register_address = (uint32_t)(address & 0x00001ffff);
+	arg.cmd52.rw_flag = (uint32_t)((direction == BUS_WRITE) ? 1 : 0);
+	arg.cmd52.write_data = value;
+	cmd.arg = arg.value;
+	WHD_BUS_STATS_INCREMENT_VARIABLE(whd_driver->bus_priv, cmd52);
+
+	result = sdhc_request(sd->sdhc, &cmd, NULL);
+
+	WHD_BUS_STATS_CONDITIONAL_INCREMENT_VARIABLE(whd_driver->bus_priv, (result != WHD_SUCCESS),
+						     cmd52_fail);
+	if (response != NULL) {
+		*response = cmd.response[0U] & SDIO_DIRECT_CMD_DATA_MASK;
+	}
+
+	/* Possibly device might not respond to this cmd. So, don't check return value here */
+	if ((result != WHD_SUCCESS) && (address == SDIO_SLEEP_CSR)) {
+		return result;
+	}
+
+	CHECK_RETURN(result);
+	return WHD_SUCCESS;
+}
+
+whd_result_t whd_bus_sdio_cmd53(whd_driver_t whd_driver, whd_bus_transfer_direction_t direction,
+				whd_bus_function_t function, sdio_transfer_mode_t mode,
+				uint32_t address, uint16_t data_size, uint8_t *data,
+				sdio_response_needed_t response_expected, uint32_t *response)
+{
+	whd_result_t result;
+	struct sd_card *sd = (struct sd_card *)whd_driver->bus_priv->sdio_obj;
+	sdio_cmd_argument_t arg = {0};
+
+	if (direction == BUS_WRITE) {
+		WHD_BUS_STATS_INCREMENT_VARIABLE(whd_driver->bus_priv, cmd53_write);
+	}
+
+	if (direction == BUS_READ) {
+		WHD_BUS_STATS_INCREMENT_VARIABLE(whd_driver->bus_priv, cmd53_read);
+	}
+
+	arg.cmd53.function_number = (uint32_t)(function & BUS_FUNCTION_MASK);
+	arg.cmd53.register_address = (uint32_t)(address & BIT_MASK(17));
```

Comment:
We only use 17 once, so we don't need a define for it. However, a comment telling the reader why 17 is the correct value to use here would be appropriate.

Reply:
we use sdio_cmd_argument_t structure to form vmd53/52 structure, so masking is not need...removed